pkg/controller/topo/entity/reconciler.go

```diff
@@ -100,31 +109,71 @@ func (r *Reconciler) reconcileCreate(entity *v1beta1.Entity) (reconcile.Result,
 		}
 	}
 
-	// Connect to the topology service
-	conn, err := grpc.ConnectService(r.client, entity.Namespace, topoService)
-	if err != nil {
-		log.Warnf("Failed to reconcile creating entity %s, %s, %s", entity.Name, entity.Namespace, err)
-		return reconcile.Result{}, err
-	}
-	defer conn.Close()
+	switch entity.Status.State {
+	case "":
+		entity.Status = v1beta1.EntityStatus{State: v1beta1.StatePending}
+		err := r.client.Status().Update(context.TODO(), entity)
+		if err != nil {
+			log.Warnf("Failed to reconcile updating state of entity %s, %s, %s", entity.Name, entity.Namespace, err)
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	case v1beta1.StatePending:
+		entity.Status = v1beta1.EntityStatus{State: v1beta1.StateAdding}
+		err := r.client.Status().Update(context.TODO(), entity)
+		if err != nil {
+			log.Warnf("Failed to reconcile updating state of entity %s, %s, %s", entity.Name, entity.Namespace, err)
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	case v1beta1.StateAdding:
+		// Check if topo service is available
+		topoNamespacedName := types.NamespacedName{Namespace: entity.Namespace, Name: topoServiceName}
+		topoServiceObject := &corev1.Service{}
+		if err := r.client.Get(context.TODO(), topoNamespacedName, topoServiceObject); err != nil && k8serrors.IsNotFound(err) {
+			// Set the state to StatePending if topo service is not found. Does it make sense????
+			entity.Status = v1beta1.EntityStatus{State: v1beta1.StatePending}
```

Comment:
Here I'm checking if topo service exists and if not then I'm updating the state to Pending. My assumption is that if topo service is not found, it's probably removed so we can remove the entities.

Reply:
Thanks for pointing this out. Actually, I think we should probably use the `Pending` state to wait for the service to exist before moving to `Adding`, that way all other states are indicative that we saw a topo service and attempted to load the entity into it, meaning if a topo service no longer exists, the entity can be safely deleted. Additionally, another possible scenario: if the topo service is deleted but the entity is not, we should move the entity back to `Pending` to wait for the topo service to be recreated (i.e. for topo to be redeployed).

Reply:
@kuujo I updated the pending state to check if topo service exists, and if not it stays pending. Please check if it's correct.